packages/rrweb/src/plugins/network/record/index.ts

```
@@ -0,0 +1,491 @@
+            ) &&
+            entry.initiatorType !== 'xmlhttprequest' &&
+            entry.initiatorType !== 'fetch'),
+      );
+    cb({
+      requests: performanceEntries.map((entry) => ({
+        url: entry.name,
+        initiatorType: entry.initiatorType as InitiatorType,
+        status: 'responseStatus' in entry ? entry.responseStatus : undefined,
+        startTime: Math.round(entry.startTime),
+        endTime: Math.round(entry.responseEnd),
+      })),
+    });
+  });
+  observer.observe({ entryTypes: ['navigation', 'resource'] });
+  return () => {
+    observer.disconnect();
+  };
+}
+
+function shouldRecordHeaders(
+  type: 'request' | 'response',
+  recordHeaders: NetworkRecordOptions['recordHeaders'],
+) {
+  return (
+    !!recordHeaders &&
+    (typeof recordHeaders === 'boolean' || recordHeaders[type])
+  );
+}
+
+function shouldRecordBody(
+  type: 'request' | 'response',
+  recordBody: NetworkRecordOptions['recordBody'],
+  headers: Headers,
+) {
+  function matchesContentType(contentTypes: string[]) {
+    const contentTypeHeader = Object.keys(headers).find(
+      (key) => key.toLowerCase() === 'content-type',
+    );
+    const contentType = contentTypeHeader && headers[contentTypeHeader];
+    return contentTypes.some((ct) => contentType?.includes(ct));
+  }
+  if (!recordBody) return false;
+  if (typeof recordBody === 'boolean') return true;
+  if (Array.isArray(recordBody)) return matchesContentType(recordBody);
+  const recordBodyType = recordBody[type];
+  if (typeof recordBodyType === 'boolean') return recordBodyType;
+  return matchesContentType(recordBodyType);
+}
+
+async function getRequestPerformanceEntry(
+  win: IWindow,
+  initiatorType: string,
+  url: string,
+  after?: number,
+  before?: number,
+  attempt = 0,
+): Promise<PerformanceResourceTiming> {
+  if (attempt > 10) {
+    throw new Error('Cannot find performance entry');
+  }
+  const urlPerformanceEntries = win.performance.getEntriesByName(
```

If I'm understanding correctly I think we'd need to make this optional... 

If people already have their own code (or another dependency) tracking performance entries they might be calling https://developer.mozilla.org/en-US/docs/Web/API/Performance/clearResourceTimings already so we'd have a race. 

Since this usage comes after the response if someone is clearing the buffer as soon as they're able to then I guess that this code would lose the race.

I think it's cheap enough to have the observer running no matter what that you could have `recordRequestPerformance` and return early here so the signature becomes `Promise<PerformanceRequestTiming | null>`

---

My mistake - 🤯 I've just read back through this and seen that the thrown error (and any other error here) are captured and swallowed where the returned promise is handled.

My preference would be to return null so that the flow is more clear but at least the error here doesn't bubble out

---

Fixed here: https://github.com/rrweb-io/rrweb/pull/1689